### PR TITLE
Update v6 core

### PR DESCRIPTION
**Pull request by danielmarbach**
Connects to #19 

The code is now at least compiling. But to be honest I'm super confused with how the Gateway is implemented. I need some help here @Particular/tf-asyncawait 

Task force: @janovesk @Scooletz

### CODE REVIEW

**Comment by SimonCropp** [comment]
@danielmarbach shouldnt the gateway actually use an incoming behaviour to get more explicit control over the pipeline?

**Comment by danielmarbach** [comment]
Probably yes. But other API provide access to ContextBag. Mutators don't.
I think we should align

-- 
*#NSBcon 2015 • The "All About NServiceBus" conference is back *
*• Learn more <http://go.particular.net/l/27442/2015-08-21/5d4wzp> →*

**Comment by SimonCropp** [comment]
>  other API provide access to ContextBag. Mutators don't. I think we should align

agreed

**Comment by andreasohlund** [comment]
After talking to @SimonCropp I'm :-1: to pass context to the mutators. I can't come up with a single use case and should the users really need it they can always use a beavior

TL;DR; 

YAGNI :)

**Comment by danielmarbach** [comment]
@SimonCropp @andreasohlund I remove the mutator. Now using two behaviors and state in-between. There are still a few todos we need to address.

**Comment by danielmarbach** [comment]
@andreasohlund can you take a look at this? Two main issues:

* Verify that the headers are correctly applied for the gateway operations
* Send to gateway satellite from the main pipeline

**Comment by janovesk** [comment]
So @Particular/tf-asyncawait what do we do with event handlers that need to be awaited?

Like https://github.com/Particular/NServiceBus.Gateway/blob/update-v6-core/src/NServiceBus.Gateway/Gateway.cs#L205 ? They return void.. which is not awesome with async/await.

**Comment by danielmarbach** [comment]
i think the design should be similar to IPushMessages. No longer use event handlers but pass in a Func<ReceivedMessage, Task> into the start method. More intention reveiling, simpler to use (no unsubscribe)...

**Comment by johnsimons** [comment]
I actually like the idea of allowing the API consumer to be fully in cpntrol of unsubscribing to the stream.
And yes IObservables to the rescue :wink:

**Comment by mauroservienti** [comment]
Callbacks are a much better design than events in this case

**Comment by janovesk** [comment]
All tests are back to green.

**Comment by janovesk** [comment]
Pretty sure we need something similar to this [async hash implementation](http://blog.cincura.net/233439-computehashasync-for-sha1/) for the MD5 calculation in https://github.com/Particular/NServiceBus.Gateway/blob/update-v6-core/src/NServiceBus.Gateway/Utils/Hasher.cs @Scooletz

**Comment by danielmarbach** [comment]
@janovesk As long as you still use https://github.com/Particular/NServiceBus.Gateway/blob/aae59b1d37da79f47720a2c07764e56340fa7296/src/NServiceBus.Gateway/Sending/SingleCallChannelForwarder.cs#L36

`ReadAsync` there is just a shim.

http://referencesource.microsoft.com/#mscorlib/system/io/memorystream.cs,393

I would shine when the core supports proper streaming for transport messages but currently not sure if it's worth

**Comment by janovesk** [comment]
This is ready for a review @Particular/tf-asyncawait.

File: src/NServiceBus.Gateway.AcceptanceTests/When_doing_request_reply.cs [hunk 1]
```diff
@@ -83,10 +87,9 @@ public SiteB()
 
             public class MyRequestHandler : IHandleMessages<MyRequest>
             {
-                public IBus Bus { get; set; }
-                public void Handle(MyRequest request)
+                public async Task Handle(MyRequest request, IMessageHandlerContext context)
                 {
-                    Bus.Return(1);
+                    await context.Reply(1);
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway.AcceptanceTests/When_doing_request_reply.cs]: You could just return the task.

File: src/NServiceBus.Gateway.AcceptanceTests/When_doing_request_reply.cs [hunk 2]
```diff
@@ -1,34 +1,38 @@
...
         {
-            var context = new Context();
-
-            Scenario.Define(context)
+            var context = await Scenario.Define<Context>()
                 .WithEndpoint<SiteA>(
-                    b => b.Given((bus, c) =>
-                        bus.SendToSites(new[] { "SiteB" }, new MyRequest())
-                            .Register(result => c.GotCallback = true)))
+                    b => b.When(async(bus, c) =>
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway.AcceptanceTests/When_doing_request_reply.cs]: Spacing in `async(bus, c)`

File: src/NServiceBus.Gateway.AcceptanceTests/When_doing_request_response_between_sites.cs [hunk 3]
```diff
@@ -83,10 +87,9 @@ public SiteB()
 
             public class MyRequestHandler : IHandleMessages<MyRequest>
             {
-                public IBus Bus { get; set; }
-                public void Handle(MyRequest request)
+                public async Task Handle(MyRequest request, IMessageHandlerContext context)
                 {
-                    Bus.Reply(new MyResponse());
+                    await context.Reply(new MyResponse());
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway.AcceptanceTests/When_doing_request_response_between_sites.cs]: You could just return the task.

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_a_message_to_another_site.cs [hunk 4]
```diff
@@ -11,12 +12,11 @@
     public class When_sending_a_message_to_another_site : NServiceBusAcceptanceTest
     {
         [Test]
-        public void Should_be_able_to_reply_to_the_message()
+        public async Task Should_be_able_to_reply_to_the_message()
         {
-            Scenario.Define<Context>()
-                    .WithEndpoint<Headquarters>(b => b.Given(bus => bus.SendToSites(new[] { "SiteA" }, new MyRequest())))
+            await Scenario.Define<Context>()
+                    .WithEndpoint<Headquarters>(b => b.When(async (bus,c) => await bus.SendToSites(new[] { "SiteA" }, new MyRequest())))
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_a_message_to_another_site.cs]: remove `async` and `await`

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_a_message_to_another_site.cs [hunk 5]
```diff
@@ -90,11 +91,10 @@ public SiteA()
 
             public class MyRequestHandler : IHandleMessages<MyRequest>
             {
-                public IBus Bus { get; set; }
 
-                public void Handle(MyRequest request)
+                public async Task Handle(MyRequest request, IMessageHandlerContext context)
                 {
-                    Bus.Reply(new MyResponse());
+                    await context.Reply(new MyResponse());
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_a_message_to_another_site.cs]: You could just return the task.

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_a_message_to_another_site.cs [hunk 6]
```diff
@@ -63,9 +63,10 @@ public class MyResponseHandler : IHandleMessages<MyResponse>
             {
                 public Context Context { get; set; }
 
-                public void Handle(MyResponse response)
+                public Task Handle(MyResponse response, IMessageHandlerContext context)
                 {
                     Context.GotResponseBack = true;
+                    return Task.FromResult(0);
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_a_message_to_another_site.cs]: Would it make sense to introduce `TaskEx.CompletedTask` as in core?

File: src/NServiceBus.Gateway.Tests/HeaderManagement/When_receiving_a_message_from_another_site.cs [hunk 7]
```diff
@@ -1,61 +1,58 @@
...
-            gatewayHeaderManager = new GatewayHeaderManager();
-
-            gatewayHeaderManager.MutateIncoming(incomingMessage);
-
-            responseMessage = new TransportMessage
-            {
-                CorrelationId = idOfIncomingMessage
-            };
-        }
+//namespace NServiceBus.Gateway.Tests.HeaderManagement
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway.Tests/HeaderManagement/When_receiving_a_message_from_another_site.cs]: Is this intended?

File: src/NServiceBus.Gateway.Tests/HeaderManagement/When_receiving_a_message_from_another_site.cs [hunk 8]
```diff
@@ -1,61 +1,58 @@
...
-        public void Should_route_the_response_to_the_replyTo_address_specified_in_the_incoming_message()
-        {
-            gatewayHeaderManager.MutateOutgoing(null, responseMessage);
-
-            Assert.AreEqual(Address.Parse(responseMessage.Headers[Headers.RouteTo]), addressOfOriginatingEndpoint);
-        }
-    }
-}
\ No newline at end of file
+    //    [Test]
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway.Tests/HeaderManagement/When_receiving_a_message_from_another_site.cs]: What about this?

File: src/NServiceBus.Gateway.Tests/Receiving/HasherTests.cs [hunk 9]
```diff
@@ -9,16 +9,16 @@ public class HasherTests
...
-        public void Valid_Md5_can_be_verified()
+        public async void Valid_Md5_can_be_verified()
         {
-            Hasher.Verify("myData".ConvertToStream(), Hasher.Hash("myData".ConvertToStream()));
+            await Hasher.Verify("myData".ConvertToStream(), "4HJGsZlkhfKtZTbdlkaTgw==");
         }
 
         [Test]
-        public void Invalid_hash_throws_ChannelException()
+        [ExpectedException(typeof(ChannelException))]
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway.Tests/Receiving/HasherTests.cs]: use `Assert.That(async() => ..., Throws...)`

File: src/NServiceBus.Gateway.Tests/Routing/When_routing_a_reply_message.cs [hunk 10]
```diff
@@ -13,17 +14,18 @@ public void Should_return_the_correct_site_based_on_the_originating_site_header(
...
             var defaultChannel = new Channel
                                      {
                                          Type = "http",
                                          Address = "http://x.y"
                                      };
 
-            message.Headers.Add(Headers.OriginatingSite, defaultChannel.ToString());
+            var headers = new Dictionary<string, string>
+            {
+                {Headers.OriginatingSite, defaultChannel.ToString()}
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway.Tests/Routing/When_routing_a_reply_message.cs]: Spacing

File: src/NServiceBus.Gateway.Tests/Routing/When_routing_using_the_configuration_source.cs [hunk 11]
```diff
@@ -22,12 +22,13 @@ public void Should_read_sites_and_their_keys_from_the_configSource()
                 Sites = section.SitesAsDictionary()
             };
 
-     
-            var message = new TransportMessage();
 
-            message.Headers.Add(Headers.DestinationSites, "SiteA");
+            var headers = new Dictionary<string, string>
+            {
+                {Headers.DestinationSites, "SiteA"}
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway.Tests/Routing/When_routing_using_the_configuration_source.cs]: Spacing

File: src/NServiceBus.Gateway/Channels/Http/HttpChannelReceiver.cs [hunk 12]
```diff
@@ -14,60 +15,115 @@ namespace NServiceBus.Gateway.Channels.Http
...
     class HttpChannelReceiver : IChannelReceiver
     {
-        public event EventHandler<DataReceivedOnChannelArgs> DataReceived;
-
-        public void Start(string address, int numberOfWorkerThreads)
+        public void Start(string address, int numberOfWorkerThreads, Func<DataReceivedOnChannelArgs, Task> dataReceivedOnChannel)
         {
+            dataReceivedHandler = dataReceivedOnChannel;
+
+            concurencyLimiter = new SemaphoreSlim(numberOfWorkerThreads, numberOfWorkerThreads);
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Channels/Http/HttpChannelReceiver.cs]: `numberOfWorkerThreads` rename to `maxConcurrency`?

File: src/NServiceBus.Gateway/Channels/Http/HttpChannelReceiver.cs [hunk 13]
```diff
@@ -14,60 +15,115 @@ namespace NServiceBus.Gateway.Channels.Http
...
             {
-                var message = string.Format("Failed to start listener for {0} make sure that you have admin privileges", address);
-                throw new Exception(message,ex);
+                var message = $"Failed to start listener for {address} make sure that you have admin privileges";
+                throw new Exception(message, ex);
             }
 
             var token = tokenSource.Token;
-            Task.Factory.StartNew(HttpServer, token, token, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+            receiverTask = Task.Factory.StartNew(state => HandleRequestAsync(state), token, token, TaskCreationOptions.LongRunning, TaskScheduler.Default).Unwrap();
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Channels/Http/HttpChannelReceiver.cs]: use `Task.Run` instead. `LongRunning` has [no or only negative effect](http://blog.i3arnon.com/2015/07/02/task-run-long-running/) with async APIs.

File: src/NServiceBus.Gateway/Channels/Http/HttpChannelReceiver.cs [hunk 14]
```diff
@@ -14,60 +15,115 @@ namespace NServiceBus.Gateway.Channels.Http
...
+            var cancellationToken = (CancellationToken)o;
+
+            while (!cancellationToken.IsCancellationRequested)
             {
-                scheduler.Dispose();
+                try
+                {
+                    var context = await listener.GetContextAsync().ConfigureAwait(false);
+
+                    var worker = Task.Run(() => Handle(context, cancellationToken), cancellationToken);
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Channels/Http/HttpChannelReceiver.cs]: I think we can remove `Task.Run` if the `Handle` only does pure async stuff

File: src/NServiceBus.Gateway/Channels/Http/HttpChannelReceiver.cs [hunk 15]
```diff
@@ -14,60 +15,115 @@ namespace NServiceBus.Gateway.Channels.Http
...
+
+                    worker.ContinueWith(t =>
+                    {
+                        Task task;
+                        runningTasks.TryRemove(worker, out task);
+                    }, cancellationToken, TaskContinuationOptions.AttachedToParent, TaskScheduler.Default)
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Channels/Http/HttpChannelReceiver.cs]: We need to remove `AttachToParent`. Doesn't work properly with async

File: src/NServiceBus.Gateway/Channels/Http/HttpChannelReceiver.cs [hunk 16]
```diff
@@ -14,60 +15,115 @@ namespace NServiceBus.Gateway.Channels.Http
...
+                    .Forget();
+
+                    runningTasks.AddOrUpdate(worker, worker, (k, v) => worker).Forget();
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Channels/Http/HttpChannelReceiver.cs]: If you use `TryAdd` you don't need `Forget`

File: src/NServiceBus.Gateway/Gateway.cs [hunk 17]
```diff
@@ -39,13 +50,25 @@ protected override void Setup(FeatureConfigurationContext context)
...
+            ConfigureRouteToGateway(context, gatewayInputAddress);
 
-            ConfigureSender(context, gatewayInputAddress);
+            context.RegisterStartupTask(b => new GatewayReceiverStartupTask(b.Build<IManageReceiveChannels>(), b.Build<IRouteMessagesToEndpoints>(), b.Build<IDispatchMessages>(), b.Build<Func<IReceiveMessagesFromSites>>(), gatewayInputAddress));
+        }
+
+        static void ConfigureRouteToGateway(FeatureConfigurationContext context, string gatewayAddress)
+        {
+            context.Pipeline.Register<RouteToGatewayBehaviour.Registration>();
+            context.Container.ConfigureComponent(b => new RouteToGatewayBehaviour(gatewayAddress), DependencyLifecycle.SingleInstance);
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Gateway.cs]: You should be able to directly new up the behavior in the registration class.

File: src/NServiceBus.Gateway/Channels/IChannelReceiver.cs [hunk 18]
```diff
@@ -8,15 +9,11 @@ namespace NServiceBus.Gateway.Channels
     public interface IChannelReceiver : IDisposable
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Channels/IChannelReceiver.cs]: Let's add a `Stop` method and check if we really need `IDisposable`

File: src/NServiceBus.Gateway/Gateway.cs [hunk 19]
```diff
@@ -125,20 +148,105 @@ static void ConfigureReceiver(FeatureConfigurationContext context, Address gatew
...
+                    new DiscardIfNotReceivedBefore(timeToBeReceived)
+                };
+
+                if(!recoverable)
+                {
+                    deliveryConstraints.Add(new NonDurableDelivery());
+                }
+
+                var operation = new UnicastTransportOperation(outgoingMessage, destination, deliveryConstraints);
+                await dispatchMessages.Dispatch(WrapInOperations(operation), new ContextBag()).ConfigureAwait(false);
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Gateway.cs]: This will be affected by https://github.com/Particular/NServiceBus/pull/3357. I'll send in a PR

File: src/NServiceBus.Gateway/HeaderManagement/GatewayIncomingBehavior.cs [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+        {
+            public string HttpFrom { get; set; }
+            public string OriginatingSite { get; set; }
+            public string ReplyToAddress { get; set; }
+            public bool LegacyMode { get; set; }
+        }
+
+        public class Registration : RegisterStep
+        {
+            public Registration() : base("GatewayIncomingBehavior", typeof(GatewayIncomingBehavior), "Extracts gateway related information from the incoming message")
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/HeaderManagement/GatewayIncomingBehavior.cs]: You can new up `GatewayIncomingBehavior ` in the registration directly. There is a func provided in the ctor

File: src/NServiceBus.Gateway/HeaderManagement/GatewayOutgoingBehavior.cs [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+            context.Headers[Headers.RouteTo] = state.ReplyToAddress;
+            // send to be backwards compatible with Gateway 3.X
+            context.Headers[GatewayHeaders.LegacyMode] = state.LegacyMode.ToString();
+
+            return next();
+        }
+
+        public class Registration : RegisterStep
+        {
+            public Registration() : base("GatewayOutgoingBehavior", typeof(GatewayOutgoingBehavior), "Puts gateway related information on the headers of outgoing messages")
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/HeaderManagement/GatewayOutgoingBehavior.cs]: You can new up `GatewayOutgoingBehavior ` in the registration directly. There is a func provided in the ctor

File: src/NServiceBus.Gateway/Receiving/IReceiveMessagesFromSites.cs [hunk 22]
```diff
@@ -1,12 +1,12 @@
...
     using System;
+    using System.Threading.Tasks;
     using Channels;
     using Notifications;
 
     interface IReceiveMessagesFromSites : IDisposable
     {
-        event EventHandler<MessageReceivedOnChannelArgs> MessageReceived;
-        void Start(Channel channel, int numberOfWorkerThreads);
+        void Start(Channel channel, int numberOfWorkerThreads, Func<MessageReceivedOnChannelArgs, Task> messageReceivedHandler);
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Receiving/IReceiveMessagesFromSites.cs]: `numberOfWorkerThreads` should be `maxConcurrency`?

File: src/NServiceBus.Gateway/Receiving/SingleCallChannelReceiver.cs [hunk 23]
```diff
@@ -214,5 +246,6 @@ static string StripSlashZeroFromCorrelationId(string corrId)
         const string Recoverable = "Recoverable";
         const string TimeToBeReceived = "TimeToBeReceived";
         static readonly TimeSpan MinimumTimeToBeReceived = TimeSpan.FromSeconds(1);
+        private Func<MessageReceivedOnChannelArgs, Task> messageReceivedHandler;
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Receiving/SingleCallChannelReceiver.cs]: Maybe we need a Syncomatic update to align the resharper style guidelines

File: src/NServiceBus.Gateway/Routing/RouteToGatewayBehaviour.cs [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+
+                context.Extensions.Remove<RouteThroughGateway>();
+            }
+
+            return next();
+        }
+
+        string gatewayAddress;
+
+        public class Registration : RegisterStep
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Routing/RouteToGatewayBehaviour.cs]: Accept in ctor here `gatewayAddress` and pass it on to the `RouteToGatewayBehaviour ` created in the func of the `RegisterStep`

File: src/NServiceBus.Gateway/Sending/GatewaySendBehavior.cs [hunk 25]
```diff
@@ -0,0 +1,117 @@
...
+        }
+
+        Task CloneAndSendLocal(byte[] body, Dictionary<string, string> headers, Site destinationSite)
+        {
+            headers[Headers.DestinationSites] = destinationSite.Key;
+
+            var message = new OutgoingMessage(headers[Headers.MessageId], headers, body);
+            var operation = new UnicastTransportOperation(message, InputAddress);
+
+            return dispatcher.Dispatch(WrapInOperations(operation), new ContextBag());
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Sending/GatewaySendBehavior.cs]: Again this will affected by https://github.com/Particular/NServiceBus/pull/3357

File: src/NServiceBus.Gateway/Sending/GatewaySendBehavior.cs [hunk 26]
```diff
@@ -0,0 +1,117 @@
...
+                throw new InvalidOperationException("No destination found for message");
+            }
+
+            await SendToSite(body, headers, destination).ConfigureAwait(false);
+        }
+
+
+        IList<Site> GetDestinationSitesFor(Dictionary<string, string> headers)
+        {
+            return builder.BuildAll<IRouteMessagesToSites>()
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Sending/GatewaySendBehavior.cs]: Could we get rid of this by adding a register method to the gateway configuration?

File: src/NServiceBus.Gateway/Utils/Hasher.cs [hunk 27]
```diff
@@ -3,26 +3,51 @@ namespace NServiceBus.Gateway.Utils
...
+        private static async Task<byte[]> ComputeHashAsync(HashAlgorithm algorithm, Stream inputStream)
+        { 
+           const int BufferSize = 4096;
+
+           algorithm.Initialize();
+
+           var buffer = new byte[BufferSize];
+           var streamLength = inputStream.Length;
+           while (true)
             {
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Utils/Hasher.cs]: Weird formatting?
